Could not compile 'pcid'

How can I solve this error?

error: borrow of packed field requires unsafe function or block (error E0133)
  --> pcid/src/main.rs:46:29
   |
46 |                             header.vendor_id, header.device_id,
   |                             ^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> pcid/src/main.rs:1:9
   |
1  | #![deny(warnings)]
   |         ^^^^^^^^
   = note: #[deny(safe_packed_borrows)] implied by #[deny(warnings)]
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>

error: borrow of packed field requires unsafe function or block (error E0133)
  --> pcid/src/main.rs:46:47
   |
46 |                             header.vendor_id, header.device_id,
   |                                               ^^^^^^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>

error: borrow of packed field requires unsafe function or block (error E0133)
  --> pcid/src/main.rs:81:33
   |
81 |                     for i in 0..header.bars.len() {
   |                                 ^^^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>

error: borrow of packed field requires unsafe function or block (error E0133)
   --> pcid/src/main.rs:127:60
    |
127 |                                         match PciBar::from(header.bars[i]) {
    |                                                            ^^^^^^^^^^^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>

error: borrow of packed field requires unsafe function or block (error E0133)
   --> pcid/src/main.rs:145:71
    |
145 |                                         "$VENID" => format!("{:>04X}",header.vendor_id),
    |                                                                       ^^^^^^^^^^^^^^^^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>

error: borrow of packed field requires unsafe function or block (error E0133)
   --> pcid/src/main.rs:146:71
    |
146 |                                         "$DEVID" => format!("{:>04X}",header.device_id),
    |                                                                       ^^^^^^^^^^^^^^^^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>

error: #[derive] can't be used on a non-Copy #[repr(packed)] struct (error E0133)
 --> pcid/src/pci/header.rs:4:10
  |
4 | #[derive(Debug, Default)]
  |          ^^^^^
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #46043 <https://github.com/rust-lang/rust/issues/46043>

error: aborting due to 7 previous errors

error: Could not compile `pcid`.
warning: build failed, waiting for other jobs to finish...
error: build failed
./repo.sh failed.mk/initfs.mk:2: recipe for target 'build/initfs.tag' failed
make: *** [build/initfs.tag] Error 1

--- a/pcid/src/main.rs
+++ b/pcid/src/main.rs
@@ -41,15 +41,13 @@ fn main() {
                 if let Some(header) = func.header() {
                     let pci_class = PciClass::from(header.class);
 
-                    let mut string;
-
-                    unsafe {
-                        string = format!("PCI {:>02X}/{:>02X}/{:>02X} {:>04X}:{:>04X} {:>02X}.{:>02X}.{:>02X}.{:>02X} {:?}",
+                    let mut string = unsafe {
+                        format!("PCI {:>02X}/{:>02X}/{:>02X} {:>04X}:{:>04X} {:>02X}.{:>02X}.{:>02X}.{:>02X} {:?}",
                                 bus.num, dev.num, func.num,
                                 header.vendor_id, header.device_id,
                                 header.class, header.subclass, header.interface, header.revision,
-                                pci_class);
-                    }
+                                pci_class)
+                    };
 
                     match pci_class {
                         PciClass::Storage => match header.subclass {

--- a/pcid/src/main.rs
+++ b/pcid/src/main.rs
@@ -38,14 +38,18 @@ fn main() {
     for bus in pci.buses() {
         for dev in bus.devs() {
             for func in dev.funcs() {
-                if let Some(header) = unsafe func.header() {
+                if let Some(header) = func.header() {
                     let pci_class = PciClass::from(header.class);
 
-                    let mut string = format!("PCI {:>02X}/{:>02X}/{:>02X} {:>04X}:{:>04X} {:>02X}.{:>02X}.{:>02X}.{:>02X} {:?}",
-                            bus.num, dev.num, func.num,
-                            header.vendor_id, header.device_id,
-                            header.class, header.subclass, header.interface, header.revision,
-                            pci_class);
+                    let mut string;
+
+                    unsafe {
+                        string = format!("PCI {:>02X}/{:>02X}/{:>02X} {:>04X}:{:>04X} {:>02X}.{:>02X}.{:>02X}.{:>02X} {:?}",
+                                bus.num, dev.num, func.num,
+                                header.vendor_id, header.device_id,
+                                header.class, header.subclass, header.interface, header.revision,
+                                pci_class);
+                    }
 
                     match pci_class {
                         PciClass::Storage => match header.subclass {
@@ -78,14 +82,16 @@ fn main() {
                         _ => ()
                     }
 
-                    for i in 0..header.bars.len() {
-                        match PciBar::from(header.bars[i]) {
-                            PciBar::None => (),
-                            PciBar::Memory(address) => string.push_str(&format!(" {}={:>08X}", i, address)),
-                            PciBar::Port(address) => string.push_str(&format!(" {}={:>04X}", i, address))
+                    unsafe {
+                        for i in 0..header.bars.len() {
+                            match PciBar::from(header.bars[i]) {
+                                PciBar::None => (),
+                                PciBar::Memory(address) => string.push_str(&format!(" {}={:>08X}", i, address)),
+                                PciBar::Port(address) => string.push_str(&format!(" {}={:>04X}", i, address))
+                            }
                         }
                     }
-
+                    
                     string.push('\n');
 
                     print!("{}", string);
@@ -124,27 +130,31 @@ fn main() {
                                 let mut command = Command::new(program);
                                 for arg in args {
                                     let bar_arg = |i| -> String {
-                                        match PciBar::from(header.bars[i]) {
-                                            PciBar::None => String::new(),
-                                            PciBar::Memory(address) => format!("{:>08X}", address),
-                                            PciBar::Port(address) => format!("{:>04X}", address)
+                                        unsafe {
+                                           match PciBar::from(header.bars[i]) {
+                                                PciBar::None => String::new(),
+                                                PciBar::Memory(address) => format!("{:>08X}", address),
+                                                PciBar::Port(address) => format!("{:>04X}", address)
+                                            } 
                                         }
                                     };
-                                    let arg = match arg.as_str() {
-                                        "$BUS" => format!("{:>02X}", bus.num),
-                                        "$DEV" => format!("{:>02X}", dev.num),
-                                        "$FUNC" => format!("{:>02X}", func.num),
-                                        "$NAME" => format!("pci-{:>02X}.{:>02X}.{:>02X}", bus.num, dev.num, func.num),
-                                        "$BAR0" => bar_arg(0),
-                                        "$BAR1" => bar_arg(1),
-                                        "$BAR2" => bar_arg(2),
-                                        "$BAR3" => bar_arg(3),
-                                        "$BAR4" => bar_arg(4),
-                                        "$BAR5" => bar_arg(5),
-                                        "$IRQ" => format!("{}", header.interrupt_line),
-                                        "$VENID" => format!("{:>04X}",header.vendor_id),
-                                        "$DEVID" => format!("{:>04X}",header.device_id),
-                                        _ => arg.clone()
+                                    let arg = unsafe {
+                                        match arg.as_str() {
+                                            "$BUS" => format!("{:>02X}", bus.num),
+                                            "$DEV" => format!("{:>02X}", dev.num),
+                                            "$FUNC" => format!("{:>02X}", func.num),
+                                            "$NAME" => format!("pci-{:>02X}.{:>02X}.{:>02X}", bus.num, dev.num, func.num),
+                                            "$BAR0" => bar_arg(0),
+                                            "$BAR1" => bar_arg(1),
+                                            "$BAR2" => bar_arg(2),
+                                            "$BAR3" => bar_arg(3),
+                                            "$BAR4" => bar_arg(4),
+                                            "$BAR5" => bar_arg(5),
+                                            "$IRQ" => format!("{}", header.interrupt_line),
+                                            "$VENID" => format!("{:>04X}",header.vendor_id),
+                                            "$DEVID" => format!("{:>04X}",header.device_id),
+                                            _ => arg.clone()
+                                        }
                                     };
                                     command.arg(&arg);
                                 }

--- a/pcid/src/main.rs
+++ b/pcid/src/main.rs
@@ -38,7 +38,7 @@ fn main() {
     for bus in pci.buses() {
         for dev in bus.devs() {
             for func in dev.funcs() {
-                if let Some(header) = func.header() {
+                if let Some(header) = unsafe func.header() {
                     let pci_class = PciClass::from(header.class);
 
                     let mut string = format!("PCI {:>02X}/{:>02X}/{:>02X} {:>04X}:{:>04X} {:>02X}.{:>02X}.{:>02X}.{:>02X} {:?}",

--- a/pcid/src/pci/header.rs
+++ b/pcid/src/pci/header.rs
@@ -1,7 +1,7 @@
 use std::ops::{Deref, DerefMut};
 use std::{slice, mem};
 
-#[derive(Debug, Default)]
+#[derive(Default)]
 #[repr(packed)]
 pub struct PciHeader {
     pub vendor_id: u16,

